pennylane/transforms/qcut.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2022 Xanadu Quantum Technologies Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module provides the circuit cutting functionality that allows large
+circuits to be distributed across multiple devices
+"""
+
+from networkx import MultiDiGraph
+
+from pennylane.measure import MeasurementProcess
+from pennylane.operation import Observable, Tensor
+from pennylane.tape import QuantumTape
+
+
+def add_operator_node(
+    graph: MultiDiGraph, op: Observable, order: int, wire_latest_node: dict
+) -> None:
+    """
+    Helper function to add operators as nodes during tape to graph connversion
+    """
+    graph.add_node(op, order=order)
+    for wire in op.wires:
+        if wire_latest_node[wire] is not None:
+            parent_op = wire_latest_node[wire]
+            graph.add_edge(parent_op, op, wire=wire)
+        wire_latest_node[wire] = op
+
+
+def tape_to_graph(tape: QuantumTape) -> MultiDiGraph:
```

Comment:
Hey @maliasadi this is the tape to graph converter mentioned in the PL meeting. It is currently using `networkx`, would be great to get your thought on a `retworkx` implementation. 

A couple of thoughts from our side:
1. We require a MultiDiGraph representation, the `rx` equivalent seems to be `PyDiGraph(...,  multigraph=True)` but perhaps there is another option?
2. It's not immediately clear how we can include our custom attributes when adding nodes to a graph in `rx`, e,g currently we have `graph.add_node(m_, order=order)`
3. What are the benefits of using `rx` over `nx`? Since we are only converting tapes to graphs the speedup that `rx` offers doesn't seem to be of much consequence but perhaps we're missing something!

Reply:
Hey @anthayes92. Thanks for sharing! Checking the code, I don't think that there is anything here `nx` offers and `rx` doesn't; however, I agree on this that the performance may be independent of the final choice.

> We require a MultiDiGraph representation, the rx equivalent seems to be PyDiGraph(..., multigraph=True) but perhaps there is another option?

You're right! The equivalence of `MultiDiGraph` is `PyDiGraph` in `rx`. You can check [`circuit_graph.py`](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/circuit_graph.py) for example.

> It's not immediately clear how we can include our custom attributes when adding nodes to a graph in rx, e,g currently we have graph.add_node(m_, order=order)

In `rx`, graph attributes can be defined as callback functions (to return statically typed data from node/edge). A good example from [here](https://qiskit.org/documentation/retworkx/networkx.html): In `nx`, we have 
```Python
import networkx as nx

graph = nx.MultiDiGraph()
graph.add_edges_from([(0, 1, {'weight': 1}), (0, 2, {'weight': 2}),
                      (1, 3, {'weight': 2}), (3, 0, {'weight': 3})])
dist_matrix = nx.floyd_warshall_numpy(graph, weight='weight')
```
while in `rx`, you may write it as,
```Python
import retworkx as rx

graph = rx.PyDiGraph()
graph.extend_from weighted_edge_list(
    [(0, 1, {'weight': 1}), (0, 2, {'weight': 2}),
     (1, 3, {'weight': 2}), (3, 0, {'weight': 3})])
dist_matrix = rx.digraph_floyd_warshall_numpy(
    graph, weight_fn=lambda edge: edge[weight])
```
> What are the benefits of using rx over nx?

As far as I dig, constructing a graph and updating nodes/edges using the `rx` Python interface shouldn't add any noticeable performance; however, when it comes to the use of graph algorithms or generators, your `rx` code will be by far faster and performant than your `nx`.  (check PR #1791 for benchmark)

Reply:
@anthayes92 I could be a bit out of the loop, but this seems interesting. Just wondering what kind of graph algorithms are we looking to run on the qcut DAGs that can benefit from `rx`, or even `nx`?

Reply:
Thanks for the useful insight @maliasadi! It looks lie this is certainly possible to do this with `rx`  but as @zeyueN has pointed out, the question is whether it is necessary. Since PL is moving toward `rx` there is some motivation there however the need for a graph structure that is well suited to graph partitioning algorithms should be the priority here. 

@zeyueN will this current conversion of tape to dag be of use for a dedicated dag class? i.e 
`tape` -> `nx.dag` -> `class DAG` 

If not then perhaps this conversion might only be used for manual circuit cutting.

Reply:
For the DAG class, I was thinking more along the lines of wrapping `nx.dag` inside it: either by making it an attribute, e.g. `self.dag`, or, directly inheriting from `MultiDiGraph`. This way, we can define our own syntax sugars for getting things like `list(g.edges(data=True))` as well as manual/auto cut functions. In other words, the `tape_to_graph` function can be the `__init__()` of the class.

But ultimately I guess this comes down to how much of these do we want to expose to users, and is having access to the vanilla `nx` graph dag important to users.

Reply:
Nice, thanks @zeyueN! It sounds like we're on a similar wavelength! Perhaps it would make sense to have the full operator decomposition as a method available in this class. 

We may include some default decomposition of specific state preparation ops and observables on the manual cutting side but not a full decomposition of all operators.

Reply:
@anthayes92 @zeyueN, a quick flyby comment, but my understanding for building the multi digraph was that two graph algorithms needed to be applied:

- Graph partitioning
- Extracting weakly connected graph fragments.

While the graph construction will likely not be any more performant using RX, my understanding is that, once the graph is constructed, applying these two graph algorithms will be much more performant in RX than NX?

Reply:
>  Though I'm cautious to put too much work into this while we have an existing CircuitGraph class that wraps nx.DiGraph (not multi). The end state might be the functionality we have here merges with the established PL CircuitGraph.

💯

Reply:
@josh146 The graph partitioning is going to be handled with a more "heavyduty" library `kahypar` (optional dependency) which uses its own graph representation. But yes the "extracting weakly connected fragments" could be more performant, especially for manually placed cuts like Tom mentioned.

Reply:
Hye @josh146, thanks for weighing in! Yes for the manual cutting case we will need an algorithm to find the weakly connected components (and possibly for the automated cutting case?) and certainly a graph partitioning algo for the automated case. 

The big question is whether using `rx` will give any advantage for graph partitioning - we have been prototyping using [kayhpar](https://github.com/kahypar/kahypar) for this fairly specialised partitioning and not using any `rx` native algos (in fact I'm not aware of any built in `rx` graph partitioning algos, though there are some built in to `nx`). 

Since kayhypar expects graphs to be of a specific format, it doesn't seem to make sense to use `rx` and also motivates creating our own kahypar-friendly DAG class (as Zeyue suggested) for the automated cutting.

Reply:
> Thanks for the useful insight @maliasadi! It looks like this is certainly possible to do this with `rx` but as @zeyueN has pointed out, the question is whether it is necessary. Since PL is moving toward `rx` there is some motivation there however the need for a graph structure that is well suited to graph partitioning algorithms should be the priority here.

I want to say if you get an up and running `nx` implementation, getting `rx` support should be quite easy as we already dug into this for `circuit_graph` and `qaoa` in PL. So, as you mentioned, we can go on with `nx` if the main use case is to construct/update graphs and talk about this more some other time. :rocket: